apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -24,131 +21,161 @@ mod public;
-
-        // Returns true if the data is deeply equal to the first argument, false
-        // otherwise. Equality is solely value-based (all JSON), no references.
-        methods.insert("eq".to_string(), future::eq_method);
-
-        // Takes any number of pairs [candidate, value], and returns value for
-        // the first candidate that equals the input data $. If none of the
-        // pairs match, a runtime error is reported, but a single-element
-        // [<default>] array as the final argument guarantees a default value.
-        methods.insert("match".to_string(), public::match_method);
-
-        // Like ->match, but expects the first element of each pair to evaluate
-        // to a boolean, returning the second element of the first pair whose
-        // first element is true. This makes providing a final catch-all case
-        // easy, since the last pair can be [true, <default>].
-        methods.insert("matchIf".to_string(), future::match_if_method);
-        methods.insert("match_if".to_string(), future::match_if_method);
-
-        // Arithmetic methods
-        methods.insert("add".to_string(), future::add_method);
-        methods.insert("sub".to_string(), future::sub_method);
-        methods.insert("mul".to_string(), future::mul_method);
-        methods.insert("div".to_string(), future::div_method);
-        methods.insert("mod".to_string(), future::mod_method);
-
-        // Array/string methods (note that ->has and ->get also work for array
-        // and string indexes)
-        methods.insert("first".to_string(), public::first_method);
-        methods.insert("last".to_string(), public::last_method);
-        methods.insert("slice".to_string(), public::slice_method);
-        methods.insert("size".to_string(), public::size_method);
-
-        // Object methods (note that ->size also works for objects)
-        methods.insert("has".to_string(), future::has_method);
-        methods.insert("get".to_string(), future::get_method);
-        methods.insert("keys".to_string(), future::keys_method);
-        methods.insert("values".to_string(), future::values_method);
-        methods.insert("entries".to_string(), public::entries_method);
-
-        // Logical methods
-        methods.insert("not".to_string(), future::not_method);
-        methods.insert("or".to_string(), future::or_method);
-        methods.insert("and".to_string(), future::and_method);
+pub(super) trait ArrowMethodImpl {
+    fn apply(
+        &self,
+        method_name: &WithRange<String>,
+        method_args: Option<&MethodArgs>,
+        data: &JSON,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
+        tail: &WithRange<PathList>,
+    ) -> (Option<JSON>, Vec<ApplyToError>);
+}
 
-        methods
-    };
+// This Deref implementation allows us to call .apply(...) directly on the
+// ArrowMethod enum.
+impl std::ops::Deref for ArrowMethod {
+    type Target = dyn ArrowMethodImpl;
```

Comment:
Oh nice, don't think I've ever done a `dyn` `Deref` before, good to know that's possible.